File: ai_agents/agents/ten_packages/extension/main_control_python/agent/agent.py
import asyncio
import json
from .llm_exec import LLMExec
from ten_runtime import AsyncTenEnv, Cmd, CmdResult, Data, StatusCode
from ten_ai_base.types import LLMToolMetadata
from .events import *

class Agent:
    def __init__(self, ten_env: AsyncTenEnv):
        self.ten_env: AsyncTenEnv = ten_env
        self.stopped = False
        self.event_queue: asyncio.Queue[AgentEvent] = asyncio.Queue()
        self.llm_exec = LLMExec(ten_env)
        self.llm_exec.on_response = self._on_llm_response  # callback handled internally

    async def on_cmd(self, cmd: Cmd):
        cmd_name = cmd.get_name()
        try:
            if cmd_name == "on_user_joined":
                event = UserJoinedEvent()
            elif cmd_name == "on_user_left":
                event = UserLeftEvent()
            elif cmd_name == "tool_register":
                tool_json, err = cmd.get_property_to_json("tool")
                if err:
                    raise RuntimeError(f"Invalid tool metadata: {err}")
                tool = LLMToolMetadata.model_validate_json(tool_json)
                event = ToolRegisterEvent(tool=tool, source=cmd.get_source().extension_name)
            else:
                self.ten_env.log_warn(f"Unhandled cmd: {cmd_name}")
                return

            await self.event_queue.put(event)
            await self.ten_env.return_result(CmdResult.create(StatusCode.OK, cmd))

        except Exception as e:
            self.ten_env.log_error(f"on_cmd error: {e}")
            await self.ten_env.return_result(CmdResult.create(StatusCode.ERROR, cmd))

    async def on_data(self, data: Data):
        data_name = data.get_name()
        try:
            if data_name == "asr_result":
                asr_json, _ = data.get_property_to_json(None)
                asr = json.loads(asr_json)
                event = ASRResultEvent(
                    text=asr.get("text", ""),
                    final=asr.get("final", False),
                    metadata=asr.get("metadata", {}),
                )
                await self.event_queue.put(event)
            else:
                self.ten_env.log_warn(f"Unhandled data: {data_name}")

        except Exception as e:
            self.ten_env.log_error(f"on_data error: {e}")

    async def get_event(self) -> AgentEvent:
        return await self.event_queue.get()


    async def register_llm_tool(self, tool: LLMToolMetadata, source: str):
        """
        Register tools with the LLM.
        This method sends a command to register the provided tools.
        """
        await self.llm_exec.register_tool(tool, source)

    async def queue_llm_input(self, text: str):
        """
        Queue a new message to the LLM context.
        This method sends the text input to the LLM for processing.
        """
        await self.llm_exec.queue_input(text)

    async def flush_llm(self):
        """
        Flush the LLM input queue.
        This will ensure that all queued inputs are processed.
        """
        await self.llm_exec.flush()

    async def stop(self):
        """
        Stop the agent processing.
        This will stop the event queue and any ongoing tasks.
        """
        self.stopped = True
        await self.llm_exec.stop()
        await self.event_queue.put(None)

    async def _on_llm_response(self, ten_env: AsyncTenEnv, delta: str, text: str, is_final: bool):
        """
        Internal callback for streaming LLM output, wrapped as an AgentEvent.
        """
        event = LLMResponseEvent(
            delta=delta,
            text=text,
            is_final=is_final,
        )
        await self.event_queue.put(event)

File: ai_agents/agents/examples/voice-assistant-nodejs/ten_packages/extension/main_nodejs/src/agent/agent.ts
//
// This file is part of TEN Framework, an open source project.
// Licensed under the Apache License, Version 2.0.
// See the LICENSE file in the root directory for more information.
//
import {
    TenEnv,
    Cmd,
    CmdResult,
    Data,
    StatusCode,
} from "ten-runtime-nodejs";
// import { LLMToolMetadata } from "ten-ai-base";
import {
    AgentEvent,
    UserJoinedEvent,
    UserLeftEvent,
    ToolRegisterEvent,
    ASRResultEvent,
    LLMResponseEvent,
} from "./events.js";
import { AsyncQueue } from "../helper.js";
import { LLMExec } from "./llm_exec.js";

// Types for handler registration
type EventHandler<T extends AgentEvent> = (event: T) => Promise<void>;

export class Agent {
    private tenEnv: TenEnv;
    private stopped = false;

    private callbacks: Map<Function, EventHandler<any>[]> = new Map();

    // Awaitable queues
    private asrQueue = new AsyncQueue<ASRResultEvent>();
    private llmQueue = new AsyncQueue<LLMResponseEvent>();

    // Drain state
    private drainingASR = false;
    private drainingLLM = false;

    // LLM task tracking / cancellation
    private llmActiveTask?: Promise<void>;
    private llmAbort = new AbortController();
    private llmExec: LLMExec;

    constructor(tenEnv: TenEnv) {
        this.tenEnv = tenEnv;
        this.llmExec = new LLMExec(
            tenEnv,
        );
        this.llmExec.onResponse = this._onLLMResponse.bind(this);
        this.llmExec.onReasoningResponse = this._onLLMReasoningResponse.bind(this);
    }

    // === Register handlers ===
    on<T extends AgentEvent>(eventClass: new (...args: any[]) => T, handler?: EventHandler<T>) {
        const register = (fn: EventHandler<T>) => {
            const list = this.callbacks.get(eventClass) || [];
            list.push(fn);
            this.callbacks.set(eventClass, list);
            return fn;
        };
        return handler ? register(handler) : register;
    }

    private async dispatch(event: AgentEvent) {
        for (const [etype, handlers] of this.callbacks.entries()) {
            if (event instanceof (etype as any)) {
                for (const h of handlers) {
                    try { await h(event); }
                    catch (err) { this.tenEnv.logError(`Handler error for ${etype}: ${err}`); }
                }
            }
        }
    }

    // === Drainers (scheduled; no polling) ===
    private scheduleASRDrain() {
        if (this.drainingASR || this.stopped) return;
        this.drainingASR = true;
        queueMicrotask(async () => {
            try {
                // drain everything currently enqueued; new arrivals will schedule another drain
                while (this.asrQueue.length > 0 && !this.stopped) {
                    const evt = await this.asrQueue.dequeue();
                    await this.dispatch(evt);
                }
            } finally {
                this.drainingASR = false;
                // If something slipped in after we checked length, schedule again.
                if (this.asrQueue.length > 0 && !this.stopped) this.scheduleASRDrain();
            }
        });
    }

    private scheduleLLMDrain() {
        if (this.drainingLLM || this.stopped) return;
        this.drainingLLM = true;
        queueMicrotask(async () => {
            try {
                // Strictly serialize LLM events
                while (this.llmQueue.length > 0 && !this.stopped) {
                    const evt = await this.llmQueue.dequeue(this.llmAbort.signal);
                    this.llmActiveTask = this.dispatch(evt);
                    try { await this.llmActiveTask; }
                    catch (e) {
                        // Abort is expected during flush() or stop()
                        if ((e as any)?.name !== 'AbortError') {
                            this.tenEnv.logError(`[Agent] LLM task failed: ${e}`);
                        }
                    } finally { this.llmActiveTask = undefined; }
                }
            } finally {
                this.drainingLLM = false;
                if (this.llmQueue.length > 0 && !this.stopped) this.scheduleLLMDrain();
            }
        });
    }

    // === Emit events ===
    private emitASR(event: ASRResultEvent) {
        this.asrQueue.enqueue(event);
        this.scheduleASRDrain();
    }

    private emitLLM(event: LLMResponseEvent) {
        this.llmQueue.enqueue(event);
        this.scheduleLLMDrain();
    }

    private async emitDirect(event: AgentEvent) {
        await this.dispatch(event);
    }

    // === Incoming from runtime ===
    async onCmd(cmd: Cmd) {
        try {
            const name = cmd.getName();
            if (name === "on_user_joined") {
                await this.emitDirect(new UserJoinedEvent());
            } else if (name === "on_user_left") {
                await this.emitDirect(new UserLeftEvent());
            } else if (name === "tool_register") {
                // parse & emit your ToolRegisterEvent here if needed
            } else {
                this.tenEnv.logWarn(`Unhandled cmd: ${name}`);
            }
            await this.tenEnv.returnResult(CmdResult.Create(StatusCode.OK, cmd));
        } catch (e) {
            this.tenEnv.logError(`onCmd error: ${e}`);
            await this.tenEnv.returnResult(CmdResult.Create(StatusCode.ERROR, cmd));
        }
    }

    async onData(data: Data) {
        try {
            if (data.getName() === "asr_result") {
                const [asrJson] = data.getPropertyToJson("");
                const asr = JSON.parse(asrJson);
                this.emitASR(new ASRResultEvent(asr.text || "", asr.final || false, asr.metadata || {}));
            } else {
                this.tenEnv.logWarn(`Unhandled data: ${data.getName()}`);
            }
        } catch (e) {
            this.tenEnv.logError(`onData error: ${e}`);
        }
    }

    queueLLMInput(input: string) {
        this.llmExec.queueInput(input);
    }

    // === LLM callbacks ===
    private async _onLLMResponse(_tenEnv: TenEnv, delta: string, text: string, isFinal: boolean) {
        this.tenEnv.logInfo(`LLM response: ${text}, isFinal: ${isFinal}, delta: ${delta}`);
        this.emitLLM(new LLMResponseEvent(delta, text, isFinal, "message"));
    }

    private async _onLLMReasoningResponse(_tenEnv: TenEnv, delta: string, text: string, isFinal: boolean) {
        this.emitLLM(new LLMResponseEvent(delta, text, isFinal, "reasoning"));
    }

    // === Optional: targeted LLM flush / cancel ===
    // You can wire this to a request_id filter if needed:
    async flushLLM() {
        await this.llmExec.flush();

        // Cancel any pending dequeue + active LLM task
        this.llmAbort.abort();
        this.llmAbort = new AbortController();   // new token for future dequeues
        this.llmQueue.clear();                   // drop queued partials
        // Note: active handler should respect AbortSignal if it does I/O
    }

    async stop() {
        this.stopped = true;
        this.flushLLM();
    }
}